modules/c++/six.sicd/source/Utilities.cpp

```diff
@@ -130,6 +130,160 @@ Utilities::getProjectionModel(const ComplexData* data,
+
+    return retv;
+}
+
+std::auto_ptr<ComplexData> six::sicd::Utilities::getComplexData(NITFReadControl& reader)
+{
+    six::Data* data = reader.getContainer()->getData(0);
+    if(data->getDataType() != six::DataType::COMPLEX)
+    {
+        throw except::Exception(Ctxt(data->getName() + " is not a SICD"));
+    }
+
+    return std::auto_ptr<ComplexData>
+        (reinterpret_cast<six::sicd::ComplexData*>(data->clone()));
+
+}
+
+std::complex<float>* six::sicd::Utilities::getWidebandData(
+        NITFReadControl& reader,
+        const ComplexData& complexData,
+        size_t bufferNumBytes,
+        std::complex<float>* buffer
+        )
+{
+    const PixelType pixelType = complexData.getPixelType();
+    const size_t imageNumber = 0;
+    const size_t startRow = 0;
+    const size_t startCol = 0;
+    const size_t numRows = complexData.getNumRows();
+    const size_t numCols = complexData.getNumCols();
+
+    std::complex<float>* returnBuffer = 0;
+    const size_t requiredBufferBytes = sizeof(std::complex<float>) 
+                                            * numRows * numCols;
+    
+
+    if(bufferNumBytes && requiredBufferBytes > bufferNumBytes)
+    {
+        // We'll take bufferNumBytes == 0 as a request for us to 
+        // allocate the buffer. If we do have bufferNumBytes, 
+        // though, make sure it's big enough. Don't do anything 
+        // sneaky where they give us a buffer but we don't use it.
+        // Throw if the given buffer isn't big enough. 
+        
+        throw except::Exception(Ctxt("Buffer provided to getWidebandData was "
+                    + str::toString(bufferNumBytes) 
+                    + " bytes, needed " 
+                    + str::toString(requiredBufferBytes) 
+                    + " bytes"));
+    }
+    else if(bufferNumBytes && !buffer)
+    {
+        // We were provided a buffer size but for some reason
+        // the buffer is still null.
+        
+        throw except::Exception(Ctxt("Null buffer provided to getWidebandData"
+                    + std::string(" when a ")
+                    + str::toString(bufferNumBytes)
+                    + std::string(" byte buffer was expected")));
+    }
+    else if(bufferNumBytes && buffer)
+    {
+        // We have buffer size AND buffer
+        returnBuffer = buffer;
+    }
+    else
+    {
+        // We do not have a buffer size, we need to make our own
+        returnBuffer = new std::complex<float>[numRows*numCols];
+    }
+
+    if(pixelType == PixelType::RE32F_IM32F)
+    {
+        six::Region region;
+        region.setStartRow(startRow);
+        region.setStartCol(startCol);
+        region.setNumRows(numRows);
+        region.setNumCols(numCols);
+        region.setBuffer(reinterpret_cast<UByte*>(returnBuffer));
+        reader.interleaved(region, imageNumber);
+    }
+    else if(pixelType == PixelType::RE16I_IM16I)
+    {
+        const size_t elementsPerRow = numCols*2;
```

Comment:
Is there a way to avoid this code duplication between the two functions where we do this reading N rows at a time and cast kind of stuff?